{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "HW2_Part_1.ipynb",
      "provenance": [],
      "collapsed_sections": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "Подготавливаем файлы для аннотации"
      ],
      "metadata": {
        "id": "5zsQwLFF5YLi"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Скачиваем скаффолды, которые мы получили в первом домашнем задании"
      ],
      "metadata": {
        "id": "rKexQiLm0o3I"
      }
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "jDYCXnmO0ZTp",
        "outputId": "b336cfb0-24f2-40a2-91bb-918cf1ced58e"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2021-12-18 18:37:17--  https://raw.githubusercontent.com/Sergey-Arakelov/hse21_hw1/main/Poil_scaffold.fa\n",
            "Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.111.133, 185.199.109.133, 185.199.110.133, ...\n",
            "Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.111.133|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 3924175 (3.7M) [text/plain]\n",
            "Saving to: ‘Poil_scaffold.fa’\n",
            "\n",
            "Poil_scaffold.fa    100%[===================>]   3.74M  --.-KB/s    in 0.09s   \n",
            "\n",
            "2021-12-18 18:37:19 (43.0 MB/s) - ‘Poil_scaffold.fa’ saved [3924175/3924175]\n",
            "\n"
          ]
        }
      ],
      "source": [
        "!wget https://raw.githubusercontent.com/Sergey-Arakelov/hse21_hw1/main/Poil_scaffold.fa"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Установка GeneMarkS-2"
      ],
      "metadata": {
        "id": "oDd2N9_T1t64"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!wget http://topaz.gatech.edu/GeneMark/tmp/GMtool_lZ6Ul/gms2_linux_64.tar.gz"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "gBJd-Sws1QO5",
        "outputId": "3a243c93-2600-4768-ebd3-581984c55fa0"
      },
      "execution_count": 3,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2021-12-18 18:39:02--  http://topaz.gatech.edu/GeneMark/tmp/GMtool_lZ6Ul/gms2_linux_64.tar.gz\n",
            "Resolving topaz.gatech.edu (topaz.gatech.edu)... 143.215.161.100\n",
            "Connecting to topaz.gatech.edu (topaz.gatech.edu)|143.215.161.100|:80... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 11261783 (11M) [application/x-gzip]\n",
            "Saving to: ‘gms2_linux_64.tar.gz’\n",
            "\n",
            "gms2_linux_64.tar.g 100%[===================>]  10.74M  2.18MB/s    in 8.5s    \n",
            "\n",
            "2021-12-18 18:39:11 (1.27 MB/s) - ‘gms2_linux_64.tar.gz’ saved [11261783/11261783]\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!wget http://topaz.gatech.edu/GeneMark/tmp/GMtool_lZ6Ul/gm_key_64.gz"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "xhNSnMYq144o",
        "outputId": "0a9ff57c-788c-44ef-d200-ae7d2cafd34b"
      },
      "execution_count": 4,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2021-12-18 18:39:58--  http://topaz.gatech.edu/GeneMark/tmp/GMtool_lZ6Ul/gm_key_64.gz\n",
            "Resolving topaz.gatech.edu (topaz.gatech.edu)... 143.215.161.100\n",
            "Connecting to topaz.gatech.edu (topaz.gatech.edu)|143.215.161.100|:80... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 86 [application/x-gzip]\n",
            "Saving to: ‘gm_key_64.gz’\n",
            "\n",
            "gm_key_64.gz        100%[===================>]      86  --.-KB/s    in 0s      \n",
            "\n",
            "2021-12-18 18:39:58 (10.7 MB/s) - ‘gm_key_64.gz’ saved [86/86]\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!gzip -d gm_key_64.gz"
      ],
      "metadata": {
        "id": "lpkyNzlX1_Us"
      },
      "execution_count": 5,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!tar -xzvf gms2_linux_64.tar.gz"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "gidDZBbf2B8r",
        "outputId": "32fec6fd-9291-4f8e-c5fd-63bb1d3151d3"
      },
      "execution_count": 6,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "gms2_linux_64/\n",
            "gms2_linux_64/gms2.pl\n",
            "gms2_linux_64/README_GMS2.txt\n",
            "gms2_linux_64/gmhmmp2_key\n",
            "gms2_linux_64/mgm_4.mod\n",
            "gms2_linux_64/gmhmmp2\n",
            "gms2_linux_64/INSTALL\n",
            "gms2_linux_64/mgm_11.mod\n",
            "gms2_linux_64/biogem\n",
            "gms2_linux_64/LICENSE\n",
            "gms2_linux_64/mgm_25.mod\n",
            "gms2_linux_64/compp\n",
            "gms2_linux_64/mgm_15.mod\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!cp  -v  gm_key_64   ~/.gmhmmp2_key"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "tj90RNwQ2DuM",
        "outputId": "b788bbfc-f374-4955-dbbd-2f2e3dd174e4"
      },
      "execution_count": 7,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "'gm_key_64' -> '/root/.gmhmmp2_key'\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Установка E-utilities (для скачивания последовательностей из NCBI)"
      ],
      "metadata": {
        "id": "NeJafMEt2JNg"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!sh -c \"$(curl -fsSL ftp://ftp.ncbi.nlm.nih.gov/entrez/entrezdirect/install-edirect.sh)\""
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "UhS53NGu2GKV",
        "outputId": "c7b42c48-d544-46f7-cd9e-896ae2893431"
      },
      "execution_count": 8,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "Entrez Direct has been successfully downloaded and installed.\n",
            "\n",
            "In order to complete the configuration process, please execute the following:\n",
            "\n",
            "  echo \"export PATH=\\${PATH}:/root/edirect\" >> ${HOME}/.bashrc\n",
            "\n",
            "or manually edit the PATH variable assignment in your .bashrc file.\n",
            "\n",
            "Would you like to do that automatically now? [y/N]\n",
            "y\n",
            "OK, done.\n",
            "\n",
            "To activate EDirect for this terminal session, please execute the following:\n",
            "\n",
            "export PATH=${PATH}:${HOME}/edirect\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!$HOME/edirect/efetch -db nuccore -id HF680312 -format gb  >  T_oleivorans_MIL_1.gbk"
      ],
      "metadata": {
        "id": "2-M2S3Wo2N9I"
      },
      "execution_count": 9,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!$HOME/edirect/efetch -db nuccore -id HF680312 -format gene_fasta  >  T_oleivorans_MIL_1.genes.fasta"
      ],
      "metadata": {
        "id": "rrUrc1gE2l0Q"
      },
      "execution_count": 10,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!$HOME/edirect/efetch -db nuccore -id HF680312 -format fasta_cds_aa  >  T_oleivorans_MIL_1.proteins.fasta"
      ],
      "metadata": {
        "id": "jV7PdX-L20N7"
      },
      "execution_count": 11,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Установка BLAST"
      ],
      "metadata": {
        "id": "3gDLo2Ob26KC"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!apt-get update"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "QhGKlPMk23Qb",
        "outputId": "e647fa5b-db51-422b-bad2-a76f2d836110"
      },
      "execution_count": 12,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\r0% [Working]\r            \rGet:1 https://cloud.r-project.org/bin/linux/ubuntu bionic-cran40/ InRelease [3,626 B]\n",
            "Ign:2 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  InRelease\n",
            "Get:3 https://cloud.r-project.org/bin/linux/ubuntu bionic-cran40/ Packages [73.9 kB]\n",
            "Ign:4 https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1804/x86_64  InRelease\n",
            "Get:5 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  Release [696 B]\n",
            "Hit:6 https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1804/x86_64  Release\n",
            "Get:7 http://ppa.launchpad.net/c2d4u.team/c2d4u4.0+/ubuntu bionic InRelease [15.9 kB]\n",
            "Get:8 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  Release.gpg [836 B]\n",
            "Get:9 http://security.ubuntu.com/ubuntu bionic-security InRelease [88.7 kB]\n",
            "Hit:11 http://archive.ubuntu.com/ubuntu bionic InRelease\n",
            "Get:12 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  Packages [833 kB]\n",
            "Get:13 http://archive.ubuntu.com/ubuntu bionic-updates InRelease [88.7 kB]\n",
            "Hit:14 http://ppa.launchpad.net/cran/libgit2/ubuntu bionic InRelease\n",
            "Get:15 http://ppa.launchpad.net/deadsnakes/ppa/ubuntu bionic InRelease [15.9 kB]\n",
            "Get:16 http://security.ubuntu.com/ubuntu bionic-security/universe amd64 Packages [1,452 kB]\n",
            "Get:17 http://archive.ubuntu.com/ubuntu bionic-backports InRelease [74.6 kB]\n",
            "Get:18 http://archive.ubuntu.com/ubuntu bionic-updates/universe amd64 Packages [2,230 kB]\n",
            "Hit:19 http://ppa.launchpad.net/graphics-drivers/ppa/ubuntu bionic InRelease\n",
            "Get:20 http://security.ubuntu.com/ubuntu bionic-security/restricted amd64 Packages [691 kB]\n",
            "Get:21 http://security.ubuntu.com/ubuntu bionic-security/main amd64 Packages [2,461 kB]\n",
            "Get:22 http://ppa.launchpad.net/c2d4u.team/c2d4u4.0+/ubuntu bionic/main Sources [1,820 kB]\n",
            "Get:23 http://archive.ubuntu.com/ubuntu bionic-updates/main amd64 Packages [2,898 kB]\n",
            "Get:24 http://archive.ubuntu.com/ubuntu bionic-backports/main amd64 Packages [11.6 kB]\n",
            "Get:25 http://archive.ubuntu.com/ubuntu bionic-backports/universe amd64 Packages [12.6 kB]\n",
            "Get:26 http://ppa.launchpad.net/c2d4u.team/c2d4u4.0+/ubuntu bionic/main amd64 Packages [933 kB]\n",
            "Get:27 http://ppa.launchpad.net/deadsnakes/ppa/ubuntu bionic/main amd64 Packages [45.9 kB]\n",
            "Fetched 13.8 MB in 7s (1,935 kB/s)\n",
            "Reading package lists... Done\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!apt-get install ncbi-blast+"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "O0a-m-IM2-6k",
        "outputId": "f131caa2-e74f-469a-84b0-9df77fa5f956"
      },
      "execution_count": 13,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Reading package lists... Done\n",
            "Building dependency tree       \n",
            "Reading state information... Done\n",
            "The following additional packages will be installed:\n",
            "  ncbi-data\n",
            "The following NEW packages will be installed:\n",
            "  ncbi-blast+ ncbi-data\n",
            "0 upgraded, 2 newly installed, 0 to remove and 79 not upgraded.\n",
            "Need to get 13.1 MB of archives.\n",
            "After this operation, 66.7 MB of additional disk space will be used.\n",
            "Get:1 http://archive.ubuntu.com/ubuntu bionic/universe amd64 ncbi-data all 6.1.20170106-2 [3,645 kB]\n",
            "Get:2 http://archive.ubuntu.com/ubuntu bionic/universe amd64 ncbi-blast+ amd64 2.6.0-1 [9,446 kB]\n",
            "Fetched 13.1 MB in 4s (3,432 kB/s)\n",
            "Selecting previously unselected package ncbi-data.\n",
            "(Reading database ... 155222 files and directories currently installed.)\n",
            "Preparing to unpack .../ncbi-data_6.1.20170106-2_all.deb ...\n",
            "Unpacking ncbi-data (6.1.20170106-2) ...\n",
            "Selecting previously unselected package ncbi-blast+.\n",
            "Preparing to unpack .../ncbi-blast+_2.6.0-1_amd64.deb ...\n",
            "Unpacking ncbi-blast+ (2.6.0-1) ...\n",
            "Setting up ncbi-data (6.1.20170106-2) ...\n",
            "Setting up ncbi-blast+ (2.6.0-1) ...\n",
            "Processing triggers for hicolor-icon-theme (0.17-2) ...\n",
            "Processing triggers for man-db (2.8.3-2ubuntu0.1) ...\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Скачиваем БД белков SwissProt"
      ],
      "metadata": {
        "id": "DwzszfwH4tob"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!wget https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_sprot.fasta.gz"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "LFXK8ZOV3jbF",
        "outputId": "de804654-e1ac-4f51-def6-c3dca0ad317c"
      },
      "execution_count": 14,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2021-12-18 18:52:04--  https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_sprot.fasta.gz\n",
            "Resolving ftp.uniprot.org (ftp.uniprot.org)... 128.175.240.195\n",
            "Connecting to ftp.uniprot.org (ftp.uniprot.org)|128.175.240.195|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 90527596 (86M) [application/x-gzip]\n",
            "Saving to: ‘uniprot_sprot.fasta.gz’\n",
            "\n",
            "uniprot_sprot.fasta 100%[===================>]  86.33M  6.11MB/s    in 19s     \n",
            "\n",
            "2021-12-18 18:52:24 (4.52 MB/s) - ‘uniprot_sprot.fasta.gz’ saved [90527596/90527596]\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!gzip -d uniprot_sprot.fasta.gz"
      ],
      "metadata": {
        "id": "3qk07M094wur"
      },
      "execution_count": 15,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Устанавливаем пакет seqkit"
      ],
      "metadata": {
        "id": "SEujOoBC46Yy"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!wget https://github.com/shenwei356/seqkit/releases/download/v2.1.0/seqkit_linux_amd64.tar.gz"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "2B3KxIgr44Ja",
        "outputId": "e8ed5ee7-f6bd-4b9a-e1e5-06f399a9d422"
      },
      "execution_count": 16,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2021-12-18 18:52:59--  https://github.com/shenwei356/seqkit/releases/download/v2.1.0/seqkit_linux_amd64.tar.gz\n",
            "Resolving github.com (github.com)... 13.114.40.48\n",
            "Connecting to github.com (github.com)|13.114.40.48|:443... connected.\n",
            "HTTP request sent, awaiting response... 302 Found\n",
            "Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/52715040/ea4018ec-70e8-40ac-a4aa-3326b04e07c4?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20211218%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20211218T185259Z&X-Amz-Expires=300&X-Amz-Signature=c6bba64645f3ed8b90ec14edd41796e7b94d6f27c9e9fefa5740176529444dc0&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=52715040&response-content-disposition=attachment%3B%20filename%3Dseqkit_linux_amd64.tar.gz&response-content-type=application%2Foctet-stream [following]\n",
            "--2021-12-18 18:52:59--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/52715040/ea4018ec-70e8-40ac-a4aa-3326b04e07c4?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20211218%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20211218T185259Z&X-Amz-Expires=300&X-Amz-Signature=c6bba64645f3ed8b90ec14edd41796e7b94d6f27c9e9fefa5740176529444dc0&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=52715040&response-content-disposition=attachment%3B%20filename%3Dseqkit_linux_amd64.tar.gz&response-content-type=application%2Foctet-stream\n",
            "Resolving objects.githubusercontent.com (objects.githubusercontent.com)... 185.199.109.133, 185.199.111.133, 185.199.110.133, ...\n",
            "Connecting to objects.githubusercontent.com (objects.githubusercontent.com)|185.199.109.133|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 6650850 (6.3M) [application/octet-stream]\n",
            "Saving to: ‘seqkit_linux_amd64.tar.gz’\n",
            "\n",
            "seqkit_linux_amd64. 100%[===================>]   6.34M  2.59MB/s    in 2.4s    \n",
            "\n",
            "2021-12-18 18:53:03 (2.59 MB/s) - ‘seqkit_linux_amd64.tar.gz’ saved [6650850/6650850]\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!tar -xzvf seqkit_linux_amd64.tar.gz"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "4WyXHaRn4-C9",
        "outputId": "a0b2f15a-5639-400b-9300-f669714c9e5b"
      },
      "execution_count": 17,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "seqkit\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!chmod a+x seqkit"
      ],
      "metadata": {
        "id": "d-J5j70I5BDC"
      },
      "execution_count": 18,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Выполнение задания"
      ],
      "metadata": {
        "id": "z0hKdL-l5RhS"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Предсказать расположение белок-кодирующих генов в геноме нашей бактерии"
      ],
      "metadata": {
        "id": "UZQ2CDAA5Qza"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Предсказать белок-кодирующие гены в собранном геноме нашей бактерии с помощью ab initio программы GeneMarkS – комманда gms2.pl По умолчанию координаты предсказанных генов сохраняются в файле gms2.lst Обратите внимание, что GeneMarkS может выдавать последовательности генов (опция --fnn), а также автоматически их транслировать в белки (--faa)"
      ],
      "metadata": {
        "id": "0ya8SQTS5lh7"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!./gms2_linux_64/gms2.pl --seq Poil_scaffold.fa --genome-type bacteria --fnn genes.fasta --faa proteins.fasta"
      ],
      "metadata": {
        "id": "clHm18RG5DXk"
      },
      "execution_count": 19,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Определяем функции белков нашей бактерии через сравнение с белками из T.oleivorans MIL-1"
      ],
      "metadata": {
        "id": "tM32mz636QB6"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Сравнить все белки (протеом) нашей бактерии с аннотированными белками T. oleivorans:\n",
        "\n",
        "Протеом T. oleivorans MIL-1 находится в файле T_oleivorans_MIL_1.proteins.fasta\n",
        "\n",
        "Индексируем белки T.oleivorans MIL-1:"
      ],
      "metadata": {
        "id": "ckB21-pS6Q4K"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!makeblastdb  -dbtype prot  -in T_oleivorans_MIL_1.proteins.fasta  -out T_oleivorans_MIL_1.proteins"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "T3CdV0lc6S4b",
        "outputId": "af61e251-9e4a-411e-8cc6-a97188a95d89"
      },
      "execution_count": 20,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "\n",
            "Building a new DB, current time: 12/18/2021 18:58:56\n",
            "New DB name:   /content/T_oleivorans_MIL_1.proteins\n",
            "New DB title:  T_oleivorans_MIL_1.proteins.fasta\n",
            "Sequence type: Protein\n",
            "Keep MBits: T\n",
            "Maximum file size: 1000000000B\n",
            "Adding sequences from FASTA; added 3666 sequences in 0.113346 seconds.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Определеям какие белки из нашей бактерии похожи на белки из T_oleivorans:"
      ],
      "metadata": {
        "id": "kM2X82XL6Wnb"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!blastp  -query proteins.fasta  -db T_oleivorans_MIL_1.proteins  -evalue 1e-10  -outfmt 6  >  scaffolds.hits_from_MIL_1.txt"
      ],
      "metadata": {
        "id": "jbuvkiU86VVE"
      },
      "execution_count": 21,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Определяем функции белков нашей бактерии через поиск по БД SwissProt"
      ],
      "metadata": {
        "id": "FsPVF-1f6jlD"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Смотрим, какие белки нашей бактерии (из Баренцева моря) не имели похожих белков из бактерии T.oleivorans MIL-1\n",
        "\n",
        "Для этого получаем список уникальных номеров генов бактерии из Барнцева моря, коротые имели хотя бы один похожий белок бактерии из Сицилли"
      ],
      "metadata": {
        "id": "Ltzt5Ic56nJ7"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!cut -f 1 scaffolds.hits_from_MIL_1.txt | sort -n | uniq > proteins.with_hits_from_MIL_1.txt"
      ],
      "metadata": {
        "id": "RjzBl-D56cuF"
      },
      "execution_count": 22,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!./seqkit grep  --invert-match  -f proteins.with_hits_from_MIL_1.txt  proteins.fasta   -o proteins.without_MIL_1.fasta"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Pp6xL5Hn6uhR",
        "outputId": "cbca308f-0257-4a55-d627-062b961ebf2f"
      },
      "execution_count": 23,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[INFO]\u001b[0m 3278 patterns loaded from file\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Проверяем, что кол-ва белков сходятся\n",
        "\n",
        "Общее кол-во белков в нашей бактерии:"
      ],
      "metadata": {
        "id": "UPWwWtbb6zer"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!grep '>' proteins.fasta | wc -l"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "2j26WjR363L7",
        "outputId": "c859b0e2-b44e-4f13-8434-16e527a98f01"
      },
      "execution_count": 24,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "3563\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Кол-во белков, которые имели схожие белки из бактерии MIL-1"
      ],
      "metadata": {
        "id": "gNv04WYP67Wb"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!wc -l proteins.with_hits_from_MIL_1.txt"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "3veukhYL6_86",
        "outputId": "b2ca5359-6e85-4e92-afca-42118b8c349b"
      },
      "execution_count": 25,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "3278 proteins.with_hits_from_MIL_1.txt\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!grep '>' proteins.without_MIL_1.fasta | wc -l"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "tYahGcEy7BpM",
        "outputId": "38f2b515-eb85-48f7-bb4e-1c9ae9f6e5f8"
      },
      "execution_count": 26,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "285\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Сравнить все белки (протеом) нашей бактерии с аннотированными белками из БД SwissProt\n",
        "\n",
        "Индексируем белки SwissProt:"
      ],
      "metadata": {
        "id": "F5V5zFki7FaS"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!makeblastdb  -dbtype prot  -in uniprot_sprot.fasta  -out uniprot_sprot"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "wUbvQYPF7Ea-",
        "outputId": "51254be7-55fd-4ba2-9b0c-6990cc5d0a96"
      },
      "execution_count": 27,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "\n",
            "Building a new DB, current time: 12/18/2021 19:02:29\n",
            "New DB name:   /content/uniprot_sprot\n",
            "New DB title:  uniprot_sprot.fasta\n",
            "Sequence type: Protein\n",
            "Keep MBits: T\n",
            "Maximum file size: 1000000000B\n",
            "Adding sequences from FASTA; added 565928 sequences in 14.5121 seconds.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Определеям какие белки из нашей бактерии похожи на белки из БД SwissProt (программа работает около 10 мин):"
      ],
      "metadata": {
        "id": "HYziLXbC7TUD"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!time  blastp  -query proteins.without_MIL_1.fasta  -db uniprot_sprot  -evalue 1e-10  -outfmt 6  >  scaffolds.hits_from_SwissProt.txt"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "02u9n70l7JNX",
        "outputId": "e91bbb82-81e6-4943-9957-5baa9314e2b6"
      },
      "execution_count": 28,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "real\t8m55.170s\n",
            "user\t8m54.300s\n",
            "sys\t0m0.336s\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Посмотрим, сколько белков удалось аннотировать через поиск по SwissProt:"
      ],
      "metadata": {
        "id": "A7sh6uu89wTE"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!cut -f 1 scaffolds.hits_from_SwissProt.txt | sort | uniq | wc -l"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "00z1s5aM7XB8",
        "outputId": "cd6ed83e-73b2-4c9c-bb09-5a91e01ae05b"
      },
      "execution_count": 29,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "52\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Загружаем файлы, которые нам потребуются для аннотации генома, на github"
      ],
      "metadata": {
        "id": "bGpMczo290bE"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "gms2.lst -- координаты всех предсказанных генов\n",
        "\n",
        "proteins.fasta -- аминокислотные последовательности всех предсказанных генов\n",
        "\n",
        "scaffolds.hits_from_MIL_1.txt -- информация о схожести белков из нашей бактерии с белками из бактерии MIL-1\n",
        "\n",
        "scaffolds.hits_from_SwissProt.txt -- информация о схожести белков из нашей бактерии с белками из БД SwissProt"
      ],
      "metadata": {
        "id": "Rdj4JBfg93PE"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Дальше пишем код на Python с использованием библиотеки Biopython и создаем геном с его аннотацией в формате GenBank (можно сделать в отдельном ноутбуке).\n",
        "\n"
      ],
      "metadata": {
        "id": "cpuk_JaK95tc"
      }
    }
  ]
}